core/qthelper.cpp

```diff
@@ -1745,3 +1745,8 @@ extern "C" void unlock_planner()
 {
 	planLock.unlock();
 }
+
+char *copy_qstring(const QString &s)
```

Comment:
i wonder if it's better to have this either as `inline` in the header or as macro. what do you think?
the compiler probably creates a branch here which then has to jump to `strdup()` in the standard library feeding the output from `qUtf8Printable()`.

Reply:
yeah, let's inline this - we use that in other super-simple functions like this

Reply:
Note that `qUtf8Printable()` itself is a macro. I'm not sure if we want to inline this [g++ 7.2 -O2]:
```
char *copy_qstring(const QString &s)
{
    5e90:       41 54                   push   %r12
    5e92:       55                      push   %rbp
    5e93:       48 89 fe                mov    %rdi,%rsi
    5e96:       53                      push   %rbx
    5e97:       48 83 ec 20             sub    $0x20,%rsp
        return strdup(qUtf8Printable(s));
    5e9b:       48 8d 5c 24 08          lea    0x8(%rsp),%rbx
    { return toUtf8_helper(*this); }
    5ea0:       48 8d 6c 24 10          lea    0x10(%rsp),%rbp
{
    5ea5:       64 48 8b 04 25 28 00    mov    %fs:0x28,%rax
    5eac:       00 00
    5eae:       48 89 44 24 18          mov    %rax,0x18(%rsp)
    5eb3:       31 c0                   xor    %eax,%eax
        return strdup(qUtf8Printable(s));
    5eb5:       48 89 df                mov    %rbx,%rdi
    5eb8:       e8 00 00 00 00          callq  5ebd <_Z12copy_qstringRK7QString+0x2d>
    5ebd:       48 89 de                mov    %rbx,%rsi
    5ec0:       48 89 ef                mov    %rbp,%rdi
    5ec3:       e8 00 00 00 00          callq  5ec8 <_Z12copy_qstringRK7QString+0x38>
    5ec8:       48 8b 7c 24 10          mov    0x10(%rsp),%rdi
    5ecd:       e8 00 00 00 00          callq  5ed2 <_Z12copy_qstringRK7QString+0x42>
    5ed2:       48 89 c7                mov    %rax,%rdi
    5ed5:       e8 00 00 00 00          callq  5eda <_Z12copy_qstringRK7QString+0x4a>
    5eda:       48 89 ef                mov    %rbp,%rdi
    5edd:       49 89 c4                mov    %rax,%r12
    5ee0:       e8 00 00 00 00          callq  5ee5 <_Z12copy_qstringRK7QString+0x55>
    5ee5:       48 89 df                mov    %rbx,%rdi
    5ee8:       e8 00 00 00 00          callq  5eed <_Z12copy_qstringRK7QString+0x5d>
}
    5eed:       48 8b 54 24 18          mov    0x18(%rsp),%rdx
    5ef2:       64 48 33 14 25 28 00    xor    %fs:0x28,%rdx
    5ef9:       00 00
    5efb:       4c 89 e0                mov    %r12,%rax
    5efe:       75 09                   jne    5f09 <_Z12copy_qstringRK7QString+0x79>
    5f00:       48 83 c4 20             add    $0x20,%rsp
    5f04:       5b                      pop    %rbx
    5f05:       5d                      pop    %rbp
    5f06:       41 5c                   pop    %r12
    5f08:       c3                      retq
    5f09:       e8 00 00 00 00          callq  5f0e <_Z12copy_qstringRK7QString+0x7e>
    5f0e:       48 89 c5                mov    %rax,%rbp
        return strdup(qUtf8Printable(s));
    5f11:       48 89 df                mov    %rbx,%rdi
    5f14:       e8 00 00 00 00          callq  5f19 <_Z12copy_qstringRK7QString+0x89>
    5f19:       48 89 ef                mov    %rbp,%rdi
    5f1c:       e8 00 00 00 00          callq  5f21 <_Z12copy_qstringRK7QString+0x91>
    5f21:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
    5f26:       66 2e 0f 1f 84 00 00    nopw   %cs:0x0(%rax,%rax,1)
    5f2d:       00 00 00
```
I don't know how much of this is function prologue and how much remains after inlining, but it doesn't look very light.

Reply:
> I don't know how much of this is function prologue and how much remains after inlining, but it doesn't look very light.

then again if you look at most QString operations they are heavy in object code too.
so should we always follow the "no-inline, no macros" policy for helpers that use QString?

to be honest i've tried it and -O2 refuses to inline it with the `inline` keyword alone, which means that it might be a bit too large. bellow you can see that the call to the branch remains and it's not inlined:
```
  40299d:	8d 5d dc             	lea    -0x24(%ebp),%ebx
  4029a0:	89 45 dc             	mov    %eax,-0x24(%ebp)
  4029a3:	89 1c 24             	mov    %ebx,(%esp)
  4029a6:	e8 25 fe ff ff       	call   4027d0 <__Z12copy_qstringRK7QString>
```

using a macro with -O2:
```
#define copy_qstring(s) \
	(char *)qUtf8Printable(s)
```

```
  402854:	c7 44 24 04 0b 00 00 	movl   $0xb,0x4(%esp)
  40285b:	00 
  40285c:	c7 04 24 6d 40 40 00 	movl   $0x40406d,(%esp)
  402863:	ff 15 88 72 40 00    	call   *0x407288
  402869:	89 45 d8             	mov    %eax,-0x28(%ebp)
  40286c:	8d 5d d8             	lea    -0x28(%ebp),%ebx
  40286f:	8d 45 dc             	lea    -0x24(%ebp),%eax
  402872:	89 5c 24 04          	mov    %ebx,0x4(%esp)
  402876:	89 04 24             	mov    %eax,(%esp)
  402879:	ff 15 84 72 40 00    	call   *0x407284
  40287f:	8b 45 dc             	mov    -0x24(%ebp),%eax
  402882:	8b 50 04             	mov    0x4(%eax),%edx
  402885:	85 d2                	test   %edx,%edx
  402887:	74 06                	je     40288f <_main+0x7f>
  402889:	83 78 0c 0f          	cmpl   $0xf,0xc(%eax)
  40288d:	76 67                	jbe    4028f6 <_main+0xe6>
  40288f:	8b 70 0c             	mov    0xc(%eax),%esi
  402892:	8b 10                	mov    (%eax),%edx
  402894:	01 c6                	add    %eax,%esi
  402896:	85 d2                	test   %edx,%edx
  402898:	74 41                	je     4028db <_main+0xcb>
  40289a:	83 c2 01             	add    $0x1,%edx
  40289d:	74 0c                	je     4028ab <_main+0x9b>
  40289f:	83 ca ff             	or     $0xffffffff,%edx
  4028a2:	f0 0f c1 10          	lock xadd %edx,(%eax)
  4028a6:	83 ea 01             	sub    $0x1,%edx
  4028a9:	74 2d                	je     4028d8 <_main+0xc8>
  4028ab:	89 d9                	mov    %ebx,%ecx
  4028ad:	e8 1e ff ff ff       	call   4027d0 <__ZN7QStringD1Ev>
  4028b2:	89 34 24             	mov    %esi,(%esp)
```

so it doesn't seem that scary in 2018. if this was 10 years ago i might have not ever inlined or used a macro for this as it's border line too big. then again how big are some of our other inlines?

final suggestion. i would add the `inline` keyword for the sake of "this looks like the right thing to do in the C++ source code` and let the compiler decide to inline or not. as you can see it refuses to inline in my case.

as a side note, looking at some of that disassembled C++ code is like looking at a pile of cr*p. but hey it works...

Reply:
> final suggestion. i would add the inline keyword

that's good but it would then has to be in header. so ignore the above and leave it the way it is.